### ISSUE TITLE
@once

### ISSUE THREAD

**Issue by chrstphrbrns** [JuliaLang/julia#51716]
It might be useful to have a `Threads.@once` macro that guarantees a line of code will only be executed once, either globally or in a given scope, a la C++ `call_once` or GCD `dispatch_once`

**Comment by nsajko** [comment]
xref [C++](https://en.cppreference.com/w/cpp/thread/call_once), [Go](https://pkg.go.dev/sync@go1.21.3#Once)

**Comment by Seelengrab** [comment]
So something like a decrement-only semaphore that other tasks aren't blocked by if they can't decrement?

**Comment by nsajko** [comment]
The go implementation actually uses a mutex:

https://github.com/golang/go/blob/be3d5fb6e6c9531e6c01dc909da710e1f596fb3a/src/sync/once.go#L18-L26

**Comment by Seelengrab** [comment]
Yes, you can use a lock and `trylock` as well. It would look like this:

```
l = Spinlock()
# ...
if trylock(l)
    # call f
end
```

**Comment by adienes** [comment]
I think this was implemented in https://github.com/JuliaLang/julia/pull/55793

I guess we can leave the issue open if someone feels like implementing a macro to use the feature more easily?

**Comment by KillerB-O** [comment]
Hi @adienes 
What is it that you guys wanted to solve here?..is the problem related mutual exclusion of a function or something else.
It Would be helpful if you could explain the problem.

**Comment by adienes** [comment]
the original request is basically solved with the `OncePerX` functions. I left the issue open to discuss potentially adding a `Threads.@once` macro-based API to maybe use the `OncePerX` functions marginally more easily

**Comment by DorkCubed** [comment]
I'd like to work on adding the macro, if possible.
To review, we are wrapping the OncePerProcess function in a Threads.@once macro, similiar to the call_once header in cpp, right?

**Comment by KillerB-O** [comment]
Okay Thanks for clarifying .
Since the core problem is already handled by the  `OncePerX` functions, now building or wrapping those functions under a `Threads.@once` macro would definitely make things much easier for users to use them.It would improve the readability of the code,lower chances of any error taking place while doing one time intialization and It also matches to what the other languages are using :`call_once` from Cpp and `sync.Once` from Go

@adienes I’d be happy to help add the macro if that’s the direction you want to go with.
